components/webext-storage/src/sync/outgoing.rs

```diff
@@ -0,0 +1,147 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+// The "outgoing" part of syncing - building the payloads to upload and
+// managing the sync state of the local DB.
+
+use interrupt::Interruptee;
+use rusqlite::{Connection, Row};
+use sql_support::ConnExt;
+use sync15::ServerTimestamp;
+use sync_guid::Guid as SyncGuid;
+
+use crate::error::*;
+
+use super::{ServerPayload, SyncStatus};
+
+// This is the "state" for outgoing items - it's so that after we POST the
+// outgoing records we can update the local DB.
+pub struct OutgoingStateHolder {
+    ext_id: String,
+    change_counter: i32,
+}
+
+pub struct OutgoingInfo {
+    state: OutgoingStateHolder,
+    payload: ServerPayload,
+}
+
+impl OutgoingInfo {
+    fn from_row(row: &Row<'_>) -> Result<Self> {
+        let guid = row
+            .get::<_, Option<SyncGuid>>("guid")?
+            .unwrap_or_else(SyncGuid::random);
+        let ext_id: String = row.get("ext_id")?;
+        let raw_data: Option<String> = row.get("data")?;
+        let (data, deleted) = if raw_data.is_some() {
+            (raw_data, false)
+        } else {
+            (None, true)
+        };
+        Ok(OutgoingInfo {
+            state: OutgoingStateHolder {
+                ext_id: ext_id.clone(),
+                change_counter: row.get("sync_change_counter")?,
```

Comment:
yes, exactly that. It could be a bool if the entire sync happened in a single transaction. Instead, we track what the value was when starting, then decrement it by that later. This doesn't have all the change counter stuff wired up (a later version is far more complete)